tests/std/include/test_atomic_wait.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#include <assert.h>
+#include <atomic>
+#include <chrono>
+#include <string.h>
+#include <thread>
+
+template <class UnderlyingType>
+void test_atomic_wait_func(const UnderlyingType old_value, const UnderlyingType new_value,
+    const std::chrono::steady_clock::duration waiting_duration) {
+    constexpr int seq_max_size = 10;
+    char seq[seq_max_size + 1];
+    std::atomic<char*> base = seq;
+    auto add_seq            = [&](char ch) {
+        char* p = base.fetch_add(1, std::memory_order_relaxed);
+        assert(p - seq < seq_max_size);
+        *p = ch;
+    };
+
+    std::atomic<UnderlyingType> a{old_value};
+    a.wait(new_value);
+
+    add_seq('1');
+
+    std::thread thd([&] {
+        std::this_thread::sleep_for(waiting_duration);
+        add_seq('2');
+        a.notify_all();
+        std::this_thread::sleep_for(waiting_duration);
+        add_seq('3');
+        a.store(old_value);
+        a.notify_one();
+        std::this_thread::sleep_for(waiting_duration);
+        add_seq('4');
+        a.store(new_value);
+        a.notify_one();
+        // timing assumption that the main thread evaluates the `wait(old_value)` before this timeout expires
+        std::this_thread::sleep_for(waiting_duration);
+        add_seq('6');
+    });
+
+    a.wait(old_value);
+    const auto loaded = a.load();
+    assert(memcmp(&loaded, &new_value, sizeof(UnderlyingType)) == 0);
+
+    add_seq('5');
+
+    thd.join();
+
+    add_seq('\0');
+    assert(strcmp(seq, "123456") == 0);
+}
+
+template <class UnderlyingType>
+void test_notify_all_notifies_all(const UnderlyingType old_value, const UnderlyingType new_value,
+    const std::chrono::steady_clock::duration waiting_duration) {
+    std::atomic<UnderlyingType> c{old_value};
+    const auto waitFn = [&c, old_value] { c.wait(old_value); };
+
+    std::thread w1{waitFn};
+    std::thread w2{waitFn};
+    std::thread w3{waitFn};
+
+    std::this_thread::sleep_for(waiting_duration);
+    c.store(new_value);
+    c.notify_all(); // if this doesn't really notify all, the following joins will deadlock
+
+    w1.join();
+    w2.join();
+    w3.join();
+}
+
+template <class UnderlyingType>
+void test_pad_bits(const std::chrono::steady_clock::duration waiting_duration) {
+    UnderlyingType old_value;
+    std::memset(&old_value, 0x66, sizeof(UnderlyingType));

```

Comment:
There are 3 calls to `std::memset` here, but you would need to include `<cstring>` for that. (If you changed from including `<string.h>`, then your unqualified calls to `memcmp`, `strcmp`, etc. would technically need to be qualified.) These should be consistent although the choice is up to you.